src/internal_telemetry/allocations/allocator/stack.rs

```diff
@@ -0,0 +1,53 @@
+use super::token::AllocationGroupId;
+
+/// An allocation group stack.
+///
+/// As allocation groups are entered and exited, they naturally end up looking a lot like a stack itself: the active
+/// allocation group gets added to the stack when entered, and if another allocation group is entered before the
+/// previous is exited, the newer group is added to the stack above the previous one, and so on and so forth.
+///
+/// This implementation is uses an array to represent the stack to avoid thread local destructor registration issues.
+#[derive(Copy, Clone)]
+pub(crate) struct GroupStack<const N: usize> {
+    slots: [AllocationGroupId; N],
+    current_top: usize,
```

Comment:
As a micro-optimization, you might get slightly better performance putting the `current_top` at the start, so the compiler doesn't have to do offset math to get at that hot element. Of course, if it's always a static, it doesn't matter…